File: Python/algorithm/src/merge_sort.py
from src.log import log


def merge(left_values, right_values):
    left_index = 0
    right_index = 0
    result = []
    while left_index < len(left_values) and right_index < len(right_values):
        if left_values[left_index] < right_values[right_index]:
            result.append(left_values[left_index])
            left_index += 1
        else:
            result.append(right_values[right_index])
            right_index += 1
    result += left_values[left_index:]
    result += right_values[right_index:]
    return result


def merge_sort(values):
    if len(values) <= 1:
        return values

    middle = len(values) // 2
    left_values = merge_sort(values[:middle])
    right_values = merge_sort(values[middle:])
    log.info("middle:%s, left:%s, right:%s" % (middle, left_values, right_values))
    return merge(left_values, right_values)

File: Go/algorithm/src/merge_sort.go
package src

func MergeSort(values []int) (result []int, err error) {
	if len(values) <= 1 {
		return values, nil
	}
	middle := len(values) >> 1
	leftValues, _ := MergeSort(values[0:middle])
	rightValues, _ := MergeSort(values[middle:])
	return Merge(leftValues, rightValues)
}

func Merge(leftValues []int, rightValues []int) ([]int, error) {
	leftIndex := 0
	rightIndex := 0
	var result []int
	for leftIndex < len(leftValues) && rightIndex < len(rightValues) {
		if leftValues[leftIndex] < rightValues[rightIndex] {
			result = append(result, leftValues[leftIndex])
			leftIndex++
		} else {
			result = append(result, rightValues[rightIndex])
			rightIndex++
		}
	}
	result = append(result, leftValues[leftIndex:]...)
	result = append(result, rightValues[rightIndex:]...)
	return result, nil
}
